src/include/mpir_request.h

```diff
@@ -392,7 +392,7 @@ MPL_STATIC_INLINE_PREFIX MPIR_Request *MPIR_Request_create_complete(MPIR_Request
     return req;
 }
 
-static inline void MPIR_Request_free(MPIR_Request * req)
+static inline void MPIR_Request_free_with_safety(MPIR_Request * req, int need_safety)
```

Comment:
I think I'd rather see us use the safe/unsafe convention vs. this API. Not sure what others think.

Reply:
Only in the `ch4` inside the device layer we have the case of device using the same lock for both operations and request pool, therefore, needs explicit interface to differentiate the cases. Outside `ch4`, there is really no case for `unsafe` version. Do you really want to expose this complications to all code paths that doesn't apply?

Reply:
Hmm, would it make sense to make this an internal `ch4` API then? I don't really want to churn ch3 callers, no.

Reply:
The whole point is not to churn `ch3` and `mpir`. They just use `MPIR_Request_free` as before.

Reply:
Right. I guess what I am saying is leave `MPIR_Request_free` as it is and add an `MPIDI_Request_free_unsafe` in ch4 only. Not sure if that is more hacky or not. Would be good to get some additional eyeballs on this commit, IMO.

Reply:
The thing is, not all calls from ch4 is `unsafe` call. My current view is to keep code outside `ch4` non-vci-aware, but inside ch4, vci-aware. So use `MPIR_Request_free_with_safety` is my intention to keep the code inside explicitly aware.

Reply:
We can't call within `ch4` like following either:
```
MPID_THREAD_CS_ENTER
MPIR_Request_free_unsafe()
MPID_THREAD_CS_EXIT
```
Because many of `MPIR_Request_free` inside ch4 doesn't really free the request, so don't need incur the critical section.

Reply:
So just going by `git grep MPIR_Request_free src/mpid/ch4`, all usages of `MPIR_Request_free_with_safety` are unsafe. There are 2 remaining usages of `MPIR_Request_free` (safe). One in `MPID_Request_free_hook` and the other in `MPIDI_workq_release_pt2pt_elemt` (untested?). Am I missing anything?

I'm not suggesting changing the `MPIR_Request_free` API to `MPIR_Request_free_safe`. Rather, leave it as-is for legacy and add an "unsafe" API for use where applicable. Or we *could* rename `MPIR_Request_free`->`MPIR_Request_free_safe` and then just macro replace it in `ch3` for compatibility.

Reply:
> There are 2 remaining usages of `MPIR_Request_free` (safe). One in `MPID_Request_free_hook` and the other in `MPIDI_workq_release_pt2pt_elemt` (untested?). Am I missing anything?

Not right now. PR #4530 is going to move the critical section from ch4-layer down to netmod/shmmod/am layer, then all ch4-layer usage will need the safe version.

Note that the safety is added deep inside `MPIR_Request_free_with_safety` when it actually need releasing the handle memory, so the safety has to be a parameter. I guess I can create another wrapper function `MPIR_Request_free_unsafe` which is `MPIR_Request_free_with_safety(req, 0)`. And maybe we can create one more wrapper `MPIR_Request_free_safe` either as macro or inline wrapper for ch4-usage. Would you prefer that?

Reply:
I initially made my comment because the "with_safety" API went against safe/unsafe convention already used in MPICH. Your point about the actual locking being within a conditional does make it harder to just simply define safe and unsafe versions. That is too bad 😦.

Reply:
That was the reason I need `with_safety` routines, but now think about it, I have no problem make that internal routines, create `free/free_safe/free_unsafe` wrappers, all wrap around the internal `with_safety` call.

Reply:
Added and used `MPIR_Reqeust_free_safe/MPIR_Request_free_unsafe` wrappers.